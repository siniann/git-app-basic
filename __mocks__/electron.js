module.exports = {
  remote: {
    dialog: {
      showSaveDialog: jest.fn(),
    },
  },
  shell: {
    openItem: jest.fn(),
  },
};
