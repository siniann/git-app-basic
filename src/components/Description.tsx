import React, { Component } from 'react';
import { TextField, Button } from '@material-ui/core';
import { descriptionRegex } from '../helper/regularExpression';

interface DescriptionState {
  description: string;
  descriptionError: string;
}

interface DescriptionProps {
  saveDescription: Function;
}
class Description extends Component<DescriptionProps, DescriptionState> {
  constructor(props) {
    super(props);
    this.state = {
      description: '',
      descriptionError: '',
    };
  }

  changeDescription = (event) => {
    const isValidSyntax = descriptionRegex(event.target.value);
    if (isValidSyntax) {
      this.setState({ descriptionError: '' });
    } else {
      this.setState({ descriptionError: 'Description should start with either Add,Correct,Change,Remove or Delete' });
    }
    if (event.target.value.length > 100) {
      this.setState({ descriptionError: 'Length should not exceed 100 characters' });
    }
    this.setState({ description: event.target.value });
  };

  saveDescription = () => {
    this.props.saveDescription(this.state.description);
  }

  render() {
    const { descriptionError, description } = this.state;


    return (
      <div>
        <br />
        <i>There are changes in the file. Please add a valid description to change to save it</i>
        <TextField
          fullWidth
          id="change-description"
          value={description}
          label="Description"
          margin="normal"
          variant="outlined"
          placeholder="Add a description..."
          onChange={this.changeDescription}
        />
        {descriptionError ? (
          <div id="description-error">
            {descriptionError}
          </div>
        )
          : null}
        <Button
          disabled={descriptionError !== '' || description === ''}
          fullWidth
          id="btn-saveDescription"
          variant="contained"
          color="default"
          onClick={this.saveDescription}
          size="small"
        >
            save description
        </Button>
      </div>
    );
  }
}

export default Description;
