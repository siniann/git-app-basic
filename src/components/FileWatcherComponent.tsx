import React, { Component } from 'react';
import Button from '@material-ui/core/Button';
import CloudUploadIcon from '@material-ui/core/SvgIcon/SvgIcon';
import { remote } from 'electron';
import TextField from '@material-ui/core/TextField';
import chokidar, { FSWatcher } from 'chokidar';
import fse from 'fs-extra';
import DefaultEditor from './DefaultEditor';
import generateTimestamp from '../helper/generateTimestamp';
import Description from './Description';
import VersionSelector from './VersionSelector';
import {
  gitAddAndCommit, isGitRepo,
} from '../helper/gitFunctions';

const path = require('path');

export interface HistoryEntry {
  version: number;
  description: string;
  timestamp: string;
  fileCopyLink: string;
}

interface FileWatcherState {
  latestVersion: number;
  filepath: string;
  history: HistoryEntry[];
  isWatching: boolean;
  showDescription: boolean;
  selectedVersion: number;

}

let watcher: FSWatcher;

class FileWatcherComponent extends Component<{}, FileWatcherState> {
  constructor(props) {
    super(props);
    this.state = {
      latestVersion: 1,
      filepath: '',
      history: [],
      isWatching: false,
      showDescription: false,
      selectedVersion: -1,
    };
  }


  onWatcherReady = () => {
    const { filepath } = this.state;
    if (fse.existsSync('./.temp')) {
      fse.removeSync('./.temp');
    }
    this.addNewEntry({
      version: this.state.latestVersion,
      description: 'Watching started',
      timestamp: generateTimestamp(),
      fileCopyLink: `./.temp/watchfolderVersions/Version${this.state.latestVersion}.txt`,
    });
    gitAddAndCommit(filepath, 'Watching started');
    console.info('Watching for changes, the initial scan completed.');
    this.setState({ isWatching: true, latestVersion: 1 });
  };

  chooseFilepath = () => {
    const openDialogReturnValuePromise = remote.dialog.showOpenDialog({ properties: ['openFile'] });
    openDialogReturnValuePromise
      .then((promiseValue) => {
        if (promiseValue.filePaths[0]) {
          const parent = path.dirname(promiseValue.filePaths[0]);
          isGitRepo(parent).then((isRepo) => {
            if (isRepo) {
              this.setState({ filepath: promiseValue.filePaths[0] });
            } else {
              this.setState({ filepath: '' });

              alert('Please choose a git repo');
            }
          });
        } else {
          console.log('No file path received');
        }
      });
  }

  startWatcher = () => {
    const {
      filepath, history, isWatching,
    } = this.state;
    if (filepath) {
      watcher = chokidar.watch(filepath, {
        ignored: /[\\]\./,
        persistent: true,
      });

      watcher
        .on('add', (pathDetail) => {
          if (isWatching) {
          // Add log
            console.log(`File added (add) : ${pathDetail}`, 'new');
          }
        })
        .on('addDir', (pathDetail) => {
          if (isWatching) {
          // Add log
            console.log(`Folder added (addDir) : ${pathDetail}`, 'new');
          }
        })
        .on('change', (pathDetail) => {
          if (isWatching) {
          // Add log
            console.log(`A change ocurred (change)  : ${pathDetail}`, 'change');
          }
        })
        .on('unlink', (pathDetail) => {
          if (isWatching) {
          // Add log
            console.log(`A file was deleted (unlink) : ${pathDetail}`, 'delete');
          }
        })
        .on('unlinkDir', (pathDetail) => {
          if (isWatching) {
          // Add log
            console.log(`A folder was deleted (unlinkdir) : ${pathDetail}`, 'delete');
          }
        })
        .on('error', (error) => {
          if (isWatching) {
            console.log('An error ocurred (error): ', 'delete');
            console.log(error);
          }
        })
        .on('ready', this.onWatcherReady)
        .on('raw', (event, path1, details) => {
          // This event should be triggered every time something happens.
          console.log('\n******-------------******\n');
          console.log('Raw event info:', event, path1, details);
          if (details.flags === 98304 && details.event === 'unknown') {
            // file opened
            if (history.length === 0) {
              alert('no entry available');
            }
          } else if (fse.existsSync('./.temp/')) {
            const { latestVersion } = this.state;

            const file1 = fse.readFileSync(`./.temp/watchfolderVersions/Version${latestVersion - 1}.txt`);
            const file2 = fse.readFileSync(filepath);
            const sameFiles = (file1.equals(file2));

            if (details.flags === 100352 && details.event === 'moved') {
              // made changes in file, but not yet saved
              if (!sameFiles) {
                this.setState({ showDescription: true });
              }
              // alert(`file up to date!${details.flags}`);
            }
            if ((details.flags === 4294656 || details.flags === 98304 || details.flags === 4295680)
                  && details.event === 'moved') {
              if (!sameFiles) {
                this.setState({ showDescription: false });
                this.addNewEntry({
                  version: latestVersion,
                  description: 'Document Updated automatically',
                  timestamp: generateTimestamp(),
                  fileCopyLink:
                        `./.temp/watchfolderVersions/Version${latestVersion}.txt`,
                });

                gitAddAndCommit(filepath, 'Document Updated automatically');
              } else {
                console.log('File is up to date.');
              }
              // console.log(this.state.history);
            }
          }
        });
    } else alert('No filepath available');
  }

  stopWatcher = () => {
    if (!watcher) {
      console.log('You need to start first the watcher');
    } else {
      watcher.close();
      // showLogFlag = false;
      this.setState({
        isWatching: false, filepath: '', history: [], showDescription: false, latestVersion: 1,
      });
      if (fse.existsSync('./.temp')) {
        fse.removeSync('./.temp');
      }
    }
  }

  filepathHandle = (event) => {
    this.setState({
      filepath: event.target.value,
    });
  }

  saveDescription = (description: string) => {
    const { filepath, latestVersion } = this.state;
    this.addNewEntry({
      version: this.state.latestVersion,
      description,
      timestamp: generateTimestamp(),
      fileCopyLink: `./.temp/watchfolderVersions/Version${latestVersion}.txt`,
    });
    this.setState({ showDescription: false });
    gitAddAndCommit(filepath, description);
  }

  dropdownChange = (selectedVersion) => {
    this.setState({
      selectedVersion,
    });
  };

  restoreVersion = () => {
    const {
      filepath, history, selectedVersion, latestVersion,
    } = this.state;
    if (selectedVersion !== -1 && selectedVersion !== undefined) {
      const versionDetailsArray = history.filter(
        (historyEntry) => String(historyEntry.version) === String(selectedVersion),
      );

      const versionDetails = versionDetailsArray[0];
      const source = versionDetails.fileCopyLink;

      try {
        fse.copySync(source, filepath);
        console.log('success replaced !');
      } catch (err) {
        console.log(err.message);
      }
      // todo: replace filepath content with the restore version content from versionDetails
      const entry = {
        version: latestVersion,
        description: `Restore version:${selectedVersion}`,
        timestamp: generateTimestamp(),
        fileCopyLink: `./.temp/watchfolderVersions/Version${latestVersion}.txt`,
      };
      this.addNewEntry(entry);
      gitAddAndCommit(filepath, entry.description);
    } else {
      alert('Please choose a version to restore!');
    }
  }


  async addNewEntry(entry: HistoryEntry) {
    const { filepath, history, latestVersion } = this.state;
    await fse.copy(filepath, `./.temp/watchfolderVersions/Version${latestVersion}.txt`)
      .then(() => {
        history.push(entry);
        const newVersion = latestVersion + 1;
        this.setState({ latestVersion: newVersion, history });
        console.log('added a new entry!');
      })
      .catch((err) => {
        console.log(err.message);
      });
  }

  render() {
    const {
      filepath, history, isWatching, showDescription, latestVersion,
    } = this.state;
    return (
      <div>
        {filepath ? (
          <TextField
            disabled
            fullWidth
            id="actual-file"
            label="Selected File"
            value={filepath}
            margin="normal"
            variant="outlined"
            onChange={this.filepathHandle}
          />
        ) : null}

        <Button
          fullWidth
          disabled={isWatching}
          id="btn-chooseFilepath"
          variant="contained"
          color="default"
          onClick={this.chooseFilepath}
        >
          Choose File
        </Button>
        {
          filepath ? (
            <div>
              <hr />
              <Button
                fullWidth
                disabled={isWatching}
                id="btn-startWatch"
                variant="contained"
                color="default"
                startIcon={<CloudUploadIcon />}
                onClick={this.startWatcher}
              >
                  Start file watching
              </Button>
              <hr />
              <Button
                fullWidth
                disabled={!isWatching}
                id="btn-stopWatch"
                variant="contained"
                color="default"
                startIcon={<CloudUploadIcon />}
                onClick={this.stopWatcher}
              >
                  Stop file watching
              </Button>
              <hr />
              <DefaultEditor
                filepath={filepath}
              />
            </div>
          )
            : null
        }
        {showDescription ? (<Description saveDescription={this.saveDescription} />) : null}

        <hr />
        {history.length === 0 ? null : (
          <div>
            <h3>
                Change History
            </h3>
            {history.slice(0).reverse().map((item) => (
              <li key={item.version}>
                {`${item.version}  | ${item.description} | ${item.timestamp}`}
              </li>
            ))}
          </div>
        )}


        {history.length > 2
          ? (
            <VersionSelector
              fileHistory={history}
              latestVersion={latestVersion - 1}
              dropdownChange={this.dropdownChange}
              restoreVersion={this.restoreVersion}
            />
          )
          : null}
      </div>

    );
  }
}
export default FileWatcherComponent;
