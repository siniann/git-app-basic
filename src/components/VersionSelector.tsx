import React, { Component, createRef } from 'react';
import Button from '@material-ui/core/Button';
import { HistoryEntry } from './FileWatcherComponent';

interface VersionSelectorProps {
  fileHistory: HistoryEntry[];
  latestVersion: number;
  dropdownChange: Function;
  restoreVersion: Function;
}

class VersionSelector extends Component<VersionSelectorProps> {
  private dropdownRef = createRef<HTMLSelectElement>();

  versionHandleChange = (event) => {
    const { dropdownChange } = this.props;
    const selectedValue = event.target.value;
    dropdownChange(selectedValue);
  };

  restore = () => {
    const { restoreVersion, dropdownChange } = this.props;
    /* istanbul ignore else */
    if (this.dropdownRef.current) {
      this.dropdownRef.current.value = `${-1}`;
      restoreVersion();
      dropdownChange(-1);
    }
  };

  render() {
    const { fileHistory, latestVersion } = this.props;
    const filteredVersions = fileHistory.filter(
      (fileHistoryEntry) => (
        fileHistoryEntry.version > 0
          && fileHistoryEntry.version !== latestVersion
      ),
    );
    const options = filteredVersions.map((fileHistoryEntry) => (
      <option
        key={fileHistoryEntry.version}
        value={fileHistoryEntry.version}
      >
        {fileHistoryEntry.version}
      </option>
    ));

    return (
      <div>
        <hr />
        Restore version
        {' '}
        <select
          id="versionDropdown"
          onChange={this.versionHandleChange}
          ref={this.dropdownRef}
        >
          <option value={-1}>Select Version</option>
          {options}
        </select>
        <br />
        <Button
          id="btn-restore"
          onClick={this.restore}
          variant="contained"
          size="small"
        >
          Restore
        </Button>

      </div>
    );
  }
}

export default VersionSelector;
