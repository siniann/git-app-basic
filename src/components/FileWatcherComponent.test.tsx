
import React from 'react';
import { shallow } from 'enzyme';
import { remote } from 'electron';
import chokidar, { FSWatcher } from 'chokidar';
import fse from 'fs-extra';

import FileWatcherComponent from './FileWatcherComponent';
import Description from './Description';
import VersionSelector from './VersionSelector';


const historyEntryArray = [
  {
    description: ' Watching started  ',
    fileCopyLink: './.temp/watchfolderVersions/Version1.txt',
    timestamp: 'Date: 2019-11-8 Time: 13:2:9',
    version: 1,
  },
  {
    description: ' update started  ',
    fileCopyLink: './.temp/watchfolderVersions/Version2.txt',
    timestamp: 'Date: 2019-11-8 Time: 13:2:10',
    version: 2,
  },
  {
    description: ' update started  ',
    fileCopyLink: './.temp/watchfolderVersions/Version3.txt',
    timestamp: 'Date: 2019-11-8 Time: 13:2:11',
    version: 3,
  },
];

const historyEntryMock = {
  description: ' Custom update  ',
  fileCopyLink: '/Users/macbook/Desktop/watchfolderVersions/Version55.txt',
  timestamp: 'Date: 2099-11-8 Time: 13:2:9',
  version: 55,
};

describe('File Watcher', () => {
  const fileWatcherShallow = shallow(<FileWatcherComponent />);
  const fileWatcherInstance = fileWatcherShallow.instance();

  beforeEach(() => {
    jest.clearAllMocks();
    window.alert = jest.fn();
    console.log = jest.fn();
    jest.mock('fs-extra');
    jest.mock('chokidar');

    jest.mock('chokidar', () => ({
      watch: jest.fn(() => ({
        on: jest.fn(function (this: any) {
          return this;
        }),
      })),
    }));

    fileWatcherInstance.setState({
      latestVersion: 1,
      filepath: '',
      history: [],
      isWatching: false,
      showDescription: false,
      selectedVersion: -1,
    });
  });

  it('should match snapshot', () => {
    expect(fileWatcherShallow).toMatchSnapshot();
    expect(fileWatcherShallow.find(Description).length).toBe(0);
    expect(fileWatcherShallow.find('#actual-file').length).toBe(0);
    expect(fileWatcherShallow.find(VersionSelector).length).toBe(0);
  });

  it('should have no filepath rendered ', () => {
    fileWatcherInstance.setState({ filepath: '' });
    expect(fileWatcherShallow).toMatchSnapshot();
    expect(fileWatcherShallow.find('#actual-file').length).toBe(0);
  });

  it('should have file path rendered', () => {
    fileWatcherInstance.setState({
      filepath: '/Users/macbook/Desktop/gitRepoCopy/Untitled.txt',
    });
    expect(fileWatcherShallow).toMatchSnapshot();
    expect(fileWatcherShallow.find('#actual-file').length).toBe(1);
  });

  it('should have description component rendered', () => {
    fileWatcherInstance.setState({
      showDescription: true,
    });
    expect(fileWatcherShallow).toMatchSnapshot();
    expect(fileWatcherShallow.find(Description).length).toBe(1);
  });


  it('should have versionSelector rendered ', () => {
    fileWatcherInstance.setState({
      history: historyEntryArray,
      isWatching: true,

    });
    expect(fileWatcherShallow.find(VersionSelector).length).toBe(1);
  });


  it('should test onWatcherReady ', async (done) => {
    await fileWatcherInstance.onWatcherReady();
    expect(fileWatcherInstance.state.isWatching).toBeTruthy();
    done();
  });

  it('should test resolved chooseFilepath', async (done) => {
    remote.dialog.showOpenDialog = jest.fn().mockResolvedValue({ filePaths: '/Users/macbook/Desktop/gitRepoCopy/Untitled.txt' });
    await fileWatcherInstance.chooseFilepath();
    expect(remote.dialog.showOpenDialog).toHaveBeenCalled();
    done();
  });


  it('should  test resolved  chooseFilepath', async (done) => {
    remote.dialog.showOpenDialog = jest.fn().mockResolvedValue({ filePaths: [] });
    await fileWatcherInstance.chooseFilepath();

    expect(remote.dialog.showOpenDialog).toHaveBeenCalled();
    expect(console.log).toHaveBeenCalledWith('No file path received');
    expect(fileWatcherInstance.state.filepath).toEqual('');
    done();
  });


  it('should  test valid restoreVersion', async (done) => {
    jest.spyOn(fse, 'copySync');
    fileWatcherInstance.setState({
      filepath: '/Users/macbook/Desktop/gitRepocopy/Untitled.txt',
      history: historyEntryArray,
      isWatching: true,
      selectedVersion: 2,
    });
    await fileWatcherInstance.restoreVersion();
    expect(fse.copySync).toHaveBeenCalled();
    done();
  });

  it('should  test invalid restoreVersion', async (done) => {
    jest.spyOn(fse, 'copySync');

    await fileWatcherInstance.restoreVersion();
    done();

    expect(window.alert).toHaveBeenCalledWith('Please choose a version to restore!');
    expect(fse.copySync).not.toHaveBeenCalled();
  });


  it('should test startWatcher with no filepath ', async (done) => {
    await fileWatcherInstance.startWatcher();
    expect(window.alert).toHaveBeenCalledWith('No filepath available');
    done();
  });

  it('should test startWatcher with filepath ', async (done) => {
    fileWatcherInstance.setState({
      filepath: '/Users/macbook/Desktop/gitRepocopy/Untitled.txt',
      isWatching: false,
    });
    fse.readFileSync = jest.fn();
    jest.spyOn(fileWatcherInstance, 'addNewEntry');

    await fileWatcherInstance.startWatcher();
    done();

    expect(fileWatcherInstance.state.filepath).toEqual('/Users/macbook/Desktop/gitRepocopy/Untitled.txt');
  });


  it('should test saveDescription with valid description ', () => {
    fileWatcherInstance.setState({
      filepath: '/Users/macbook/Desktop/gitRepocopy/Untitled.txt',
      history: historyEntryArray,
      isWatching: true,
      showDescription: true,
      latestVersion: 3,

    });
    jest.spyOn(fileWatcherInstance, 'addNewEntry');
    fileWatcherInstance.saveDescription('Add valid des');
    expect(fileWatcherInstance.addNewEntry).toHaveBeenCalled();
    expect(fileWatcherInstance.state.showDescription).toEqual(false);
  });


  it('should test dropdownChange with valid input ', () => {
    fileWatcherInstance.dropdownChange(4);

    expect(fileWatcherInstance.state.selectedVersion).toEqual(4);
  });

  it('should test filepathHandle with event input', () => {
    const eventMock = {
      target: { value: 'some file path' },
    };
    fileWatcherInstance.filepathHandle(eventMock);

    expect(fileWatcherInstance.state.filepath).toBe('some file path');
  });


  it('should test resolved addNewEntry with valid input historyEntry ', async (done) => {
    fileWatcherInstance.setState({
      history: historyEntryArray,
      isWatching: true,
      latestVersion: 3,

    });
    fse.copy = jest.fn().mockResolvedValue({});
    await fileWatcherInstance.addNewEntry(historyEntryMock).then(() => {
      expect(console.log).toHaveBeenCalledWith('added a new entry!');
      expect(fileWatcherInstance.state.latestVersion).toBe(4);
    });

    done();
  });

  it('should test addNewEntry rejected promise ', async (done) => {
    fse.copy = jest.fn().mockRejectedValue({ message: 'error' });
    fileWatcherInstance.setState({
      history: historyEntryArray,
      isWatching: true,
      latestVersion: 3,

    });
    await fileWatcherInstance.addNewEntry(historyEntryMock).then(() => {
      expect(fileWatcherInstance.state.latestVersion).toBe(3);
    });
    done();
  });
});
