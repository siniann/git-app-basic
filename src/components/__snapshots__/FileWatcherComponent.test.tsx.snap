// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Watcher should have description component rendered 1`] = `ShallowWrapper {}`;

exports[`File Watcher should have file path rendered 1`] = `ShallowWrapper {}`;

exports[`File Watcher should have no filepath rendered  1`] = `ShallowWrapper {}`;

exports[`File Watcher should match snapshot 1`] = `ShallowWrapper {}`;
